Fix potential READ memory access

vrabaud opened the pull request:
This fixes https://oss-fuzz.com/testcase-detail/4923671881252864 and https://oss-fuzz.com/testcase-detail/5048650127966208

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

vrabaud commented:
@asmorkalov, I see that tests are failing due to old GitHub dependencies. We probably want a dependabot config. This is what we have for AVIF: https://github.com/AOMediaCodec/libavif/blob/main/.github/dependabot.yml

shyama7004 commented:
> @asmorkalov, I see that tests are failing due to old GitHub dependencies. We probably want a dependabot config. This is what we have for AVIF: https://github.com/AOMediaCodec/libavif/blob/main/.github/dependabot.yml
> 
> It will make PRs once a month to update your CI files (if needed)

Due to  the deprecation of actions/upload-artifact@v3, upgrading to v4.6.0 can be a good move. It improves performance with faster uploads and immediate artifact availability.

warped-rudi commented:
@vrabaud Can you elaborate on this a bit more? IMHO there is nothing wrong with the original code.

asmorkalov commented:
@vrabaud Could you bring more details. I do not see any difference.

warped-rudi commented:
There is another small thing I find unfortunate: The default constructor of 'PngPtrs' allocates data. That means, when a PngDecoder is instantiated, we allocate stuff that gets thrown away on the very first usage of this decoder instance. I guess it's better to add an explicit 'init' method to 'PngPtrs' and let the default constructor just zero-initialize all member variables.

vrabaud commented:
@warped-rudi , I indeed refactored to be more consistent. The READ memory access is fixed by the extra `setjmp` in readData`.

vrabaud commented:
I added a fix for https://oss-fuzz.com/testcase-detail/5048650127966208

warped-rudi commented:
That look good to me! Two more nit-picks: PngDecoder doesn't have to have protected members. Make them private. Also the question wether or not to call ClearPngPtr() every time setjmp() returns a non-zero value persists. Currently it's only done in processing_finish().

asmorkalov commented:
Linux builds report:
```
/home/ci/opencv/modules/imgcodecs/src/grfmt_png.cpp:540:21: warning: variable 'buffer' might be clobbered by 'longjmp' or 'vfork' [-Wclobbered]
```

vrabaud commented:
I fixed the two comments. 

> That look good to me! Two more nit-picks: PngDecoder (as well as PngEncoder) don't have to have protected members. Make them private. Also the question wether or not to call ClearPngPtr() every time setjmp() returns a non-zero value persists. Currently it's only done in processing_finish().

ClearPngPtr does not need to be called every time setjmp returns a non-zero as it deals with variables outside of the longjmp. Those variables will be cleared normally by the destructor. I removed it from processing_finish.
